src/shared/sol-iio.c

```diff
@@ -0,0 +1,990 @@
+            device->device_id);
+        return false;
+    }
+
+    len = snprintf(path, sizeof(path), CHANNEL_RAW_PATH, device->device_id,
+        channel->name);
+    if (len < 0) {
+        SOL_WRN("Could not read channel [%s] in device%d", channel->name,
+            device->device_id);
+        return false;
+    }
+
+    len = sol_util_read_file(path, "%d", &raw_value);
+    if (len < 0) {
+        SOL_WRN("Could not read channel [%s] in device%d", channel->name,
+            device->device_id);
+        return false;
+    }
+
+    *value = (raw_value + channel->offset) * channel->scale;
+    return true;
+}
+
+void
+sol_iio_set_reader_cb(struct sol_iio_device *device, sol_iio_reader_cb reader_cb, void *data)
+{
+    device->reader_cb = reader_cb;
+    device->reader_cb_data = data;
+}
+
+static int
+calc_channel_offset_in_buffer(struct sol_iio_channel *channel)
+{
+    struct sol_iio_channel *channel_itr;
+    int i, offset = 0;
+
+    SOL_PTR_VECTOR_FOREACH_IDX (&channel->device->channels, channel_itr, i) {
+        if (channel_itr->index < channel->index)
+            offset += channel->storagebits;
+    }
+
+    return offset;
+}
+
+bool
+sol_iio_read_buffer_channel_value(struct sol_iio_channel *channel, struct sol_iio_buffer *buffer, double *value)
+{
+    uint64_t data = 0;
+    int64_t s_data = -1;
+    int i, j, offset_bytes, storage_bytes;
+    bool negative = false;
+
+    /* Leve this calc to the first read, as now we know how much channels are enabled */
+    if (channel->offset_in_buffer == -1)
+        channel->offset_in_buffer = calc_channel_offset_in_buffer(channel);
+
+    if (channel->offset_in_buffer + channel->storagebits > (int)buffer->size * 8) {
+        SOL_WRN("Invalid read on buffer - is buffer complete?");
+        return false;
+    }
+
+    if (channel->storagebits > 64) {
+        // XXX is such a thing even possible?
+        SOL_WRN("Could not read channel - more than 64 bits");
+        return false;
+    }
+
+    offset_bytes = channel->offset_in_buffer / 8;
+    storage_bytes = channel->storagebits / 8;
+    if (channel->little_endian) {
+        for (i = 0, j = 0; i < storage_bytes; i++, j += 8) {
+            data |= buffer->data[i + offset_bytes] << j;
+        }
+    } else {
+        for (i = storage_bytes - 1, j = 0; i >= 0; i--, j += 8) {
+            data |= buffer->data[i + offset_bytes] << j;
+        }
+    }
+
+    data >>= channel->shift;
+
+    /* Remove the top useless bits */
+    data <<= channel->storagebits - channel->bits;
+    data >>= channel->storagebits - channel->bits;
+
+    /* If signed and msb is 1, we have a negative number */
+    if (channel->is_signed) {
+        negative = data & (1 << (channel->bits - 1));
+    }
+
+    if (negative) {
+        s_data = data | (s_data << channel->bits);
+        *value = (s_data + channel->offset) * channel->scale;
+    } else
+        *value = (data + channel->offset) * channel->scale;
```

Comment:
As far as I tested, it doesn't work. How would it convert if data is an `uint64_t`? The bit pattern on the last n bits, say `0000 ... 1111 1111` would represent `-1` as an `int8_t`, but is `255` on the `uint64_t`; if multiplied by `-1`, yields to `-255`.